### DISCUSSION TITLE
Issue with Re-merging Shapes in diagrams.net Embedded in a Web Application

### DISCUSSION THREAD

**Discussion by reinoutkoninkx** [jgraph/drawio#4229] [Q&A]
I am currently working with a web application that incorporates an embedded version of diagrams.net within an iframe. We manage the hosting of diagrams directly in our own database, and this setup has been functioning smoothly.

In our application, we provide users with a selection of custom shapes that are intended for use within their diagrams. These shapes are accessible through a custom menu. To integrate these shapes into the diagrams, we utilize the PostMessage API with the 'merge' action. This method allows for the shapes to be seamlessly integrated into the diagrams, maintaining the focus without the disruptions typically caused by the 'load' action. This process has been working exceptionally well.

However, we have encountered an issue when a shape, previously added to a diagram, is deleted. Although the shape appears to be removed successfully from the diagram, any attempt to re-merge the same shape (using the identical id) into the diagram fails; the shape does not get re-added. This issue persists until the browser window is refreshed. After a refresh, the shape can be merged again, but the problem reoccurs if the shape is removed and an attempt is made to merge it back in.

Has anyone experienced a similar issue or can provide insights into what might be causing this behavior and how to resolve it?

I look forward to any help!

Thanks,
Reinout

**Comment by davidjgraph** [comment]
How is the re-added shape using the old ID? Is it an undo?

**Reply by reinoutkoninkx** [reply]
Not an undo, the custom menu we have is outside the iframe, it uses the PostMessage API and another custom API to create its XML which also works well. 

By clicking on a shape in our own menu its xml is sent into the diagram with the same id it had before when it got removed, this then does not get added into the diagram because it seems to think a shape with the same id is already in the diagram.

**Reply by davidjgraph** [reply]
Hard to say, is there a reproduction case accessible publicly?

**Reply by reinoutkoninkx** [reply]
I can make a video for you where I do it? Will that suffice? I can not give you the code however.

**Reply by davidjgraph** [reply]
No, we need a reproduction case.

**Reply by reinoutkoninkx** [reply]
I have a testenvironment that you can use if I assist you through the login procedure? and then redirect you to how you can reproduce this said issue?

**Reply by reinoutkoninkx** [reply]
I think the general issue might be that you keep track of ids in the diagram based on certain actions (one of which you mentioned, an undo or redo), but I don't think you update the list of ids in the diagram when using the merge action in using the PostMessage API.

**Reply by reinoutkoninkx** [reply]
Have you got an update, how can we proceed?
